master/pyext/src/validation/__init__.py

```diff
@@ -12,527 +12,623 @@
-        if len(self.system.orphan_datasets)>0:
+        dataset_dic = defaultdict()
+        if len(self.system.orphan_datasets) > 0:
             for i in self.system.orphan_datasets:
-                if i.data_type =='unspecified':
-                    dataset_dic[str(i._id)]=str(i.details)
+                if i.data_type == 'other':
+                    dataset_dic[str(i._id)] = str(i.details)
                 else:
-                    dataset_dic[str(i._id)]=str(i.data_type)
+                    try:
+                        dataset_dic[str(i._id)] = str(i.data_type)
+                    except TypeError:
+                        dataset_dic[str(i._id)] = 'None'
         return dataset_dic
-        
-    def get_restraints(self)->dict:
+
+    def get_restraints(self) -> dict:
         """ get restraints table from cif file"""
-        r=self.system.restraints
-        restraints_comp={'ID':[],'Dataset ID':[],'Restraint type':[],'Restraint info':[]}
-        for j,i in enumerate(r):
+        r = self.system.restraints
+        restraints_comp = {'ID': [], 'Dataset ID': [],
+                           'Restraint type': [], 'Restraint info': []}
+        for j, i in enumerate(r):
             restraints_comp['ID'].append(j+1)
             restraints_comp['Restraint type'].append(str(i.__class__.__name__))
-            restraints_comp['Dataset ID'].append(str(i.dataset._id))
-            if 'CrossLink' in str(i.__class__.__name__):
-                restraints_comp['Restraint info'].append(str(i.linker.auth_name)+', '+str(len(i.experimental_cross_links))+' cross-links')
-            if 'EM3D' in str(i.__class__.__name__):
-                restraints_comp['Restraint info'].append(str(i.fitting_method)+ ', '+str(i.number_of_gaussians))
-            if 'EM2D' in str(i.__class__.__name__):
-                restraints_comp['Restraint info'].append('Number of micrographs: '+str(i.number_raw_micrographs)+','+' Image resolution: '+str(i.image_resolution))
-            if 'SAS' in str(i.__class__.__name__):
-                restraints_comp['Restraint info'].append('Assembly name: '+str(i.assembly.name)+' Fitting method: '+str(i.fitting_method)+ ' Multi-state: '+str(i.multi_state))
-            if 'UpperBound' in str(i.__class__.__name__):
-
-                restraints_comp['Restraint info'].append('Distance: '+str(i.distance))
+            try:
+                restraints_comp['Dataset ID'].append(str(i.dataset._id))
+            except AttributeError:
+                restraints_comp['Dataset ID'].append('None')
+            # print (i.__class__.__name__,i,i.__class__,type(i))
+            # print (isinstance(i,ihm.restraint.CrossLinkRestraint))
+            if isinstance(i, ihm.restraint.CrossLinkRestraint):
+                restraints_comp['Restraint info'].append(
+                    str(i.linker.auth_name) + ', ' +
+                    str(len(i.experimental_cross_links)) + ' cross-links')
+            if isinstance(i, ihm.restraint.EM3DRestraint):
```

Comment:
These should probably be `elif`s since this code currently implies that a given restraint might be an instance of multiple base classes.